### PR TITLE
table component improvement

### PR DESCRIPTION
**Pull request by alanrenmsft**
migration experience needs button and hyperlink column types in table component.

1. new hyperlink column for table component.
1. code refactoring, move the clickable column logic to a base class and let button column and hyperlink column inherit it.
1. update sqlservices sample extension to use the new features.

image button, text button, link with icon, link
image

UI impacted by this PR, no visual changes:

1. Schema Compare Main
before

image

after
image

1. Schema Compare options dialog
before
image

after
image

1. SQL Assesment
before
image

after
image

1. Database list
before
image

after
image

### CODE REVIEW

File: src/sql/azdata.d.ts [hunk 1]
```diff
@@ -3222,6 +3222,9 @@ declare module 'azdata' {
 		headerCssClass?: string;
 		toolTip?: string;
 		type?: ColumnType;
+		/**
+		 * @deprecated options property is deprecated, use specific column types to access the options directly
+		 */
 		options?: CheckboxColumnOption | TextColumnOption;
```
Comment (alanrenmsft) [inline src/sql/azdata.d.ts:3228]: I am not able to update the options to include new types without updating azdata.d.ts directly. using more specific column types can solve this problem.

File: src/sql/azdata.proposed.d.ts [hunk 2]
```diff
@@ -842,21 +842,76 @@ declare module 'azdata' {
 	}
 
 	export interface IconColumnCellValue {
+		/**
+		 * The icon to be displayed.
+		 */
 		icon: string | vscode.Uri | { light: string | vscode.Uri; dark: string | vscode.Uri };
-		ariaLabel: string;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts:846]: update it to title to be more generic.

Comment (alanrenmsft) [reply src/sql/azdata.proposed.d.ts:846]: title is also using the same value

File: src/sql/base/browser/ui/table/plugins/tableColumn.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+export interface TableCellClickEventArgs<T extends Slick.SlickData> {
+	item: T;
+	position: { x: number, y: number };
+	row: number;
+	column: number;
+}
+
+export abstract class BaseClickableColumn<T extends Slick.SlickData> implements Slick.Plugin<T>, TableColumn<T> {
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/plugins/tableColumn.ts:20]: The code here is from the old buttonColumn definition, moved it here for code reuse.

File: src/sql/workbench/browser/modelComponents/table.component.ts [hunk 4]
```diff
@@ -382,15 +426,19 @@ export default class TableComponent extends ComponentBase<azdata.TableComponentP
 		}
 	}
 
-	private createButtonPlugin(col: any) {
+	private createButtonPlugin(col: azdata.ButtonColumn) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/browser/modelComponents/table.component.ts:429]: Seems kind of weird to call these `createXPlugin` when they aren't really creating a plugin - just a column

**Review by Charles-Gagnon — APPROVED** [review]

File: src/sql/workbench/browser/modelComponents/table.component.ts (hunk 4 above)
Comment (alanrenmsft) [reply src/sql/workbench/browser/modelComponents/table.component.ts:429]: these are actually plugins as well. :)

image

**Comment by alanrenmsft** [comment]
noticed the width of  'more actions' column has been changed while comparing screenshots, update the code to use fixed width for image only button column to match the previous behavior.

**Comment by alanrenmsft** [comment]
before:
image

after
image

**Pull request merged**